backend/manager/modules/common/src/main/java/org/ovirt/engine/core/common/businessentities/VM.java

```diff
@@ -2008,4 +2008,20 @@ public CpuPinningPolicy getCpuPinningPolicy() {
     public void setCpuPinningPolicy(CpuPinningPolicy cpuPinningPolicy) {
         vmStatic.setCpuPinningPolicy(cpuPinningPolicy);
     }
+
+    /**
+     * Get the CPU pinning string for the VM.
+     *
+     * Manual policy is an exception in which the pinning is specified in vm_static, otherwise
+     * we take it from vm_dynamic.
+     * @return the CPU pinning string.
+     */
+    public String getVmPinning() {
+        if (getCpuPinningPolicy() == CpuPinningPolicy.MANUAL) {
```

Comment:
The `getCpuPinningPolicy()` here returns only the policy configured by user? Or how do you distinguish it from the case where the policy is overridden in VMHandler.java so that you can fall through to the lines below?

Reply:
That depends on the VM instance.
But let's say on the above discussion (and now when I looked on the code from that line of calling VmHandler to sending the create API to VDSM):
There are two places we look into the policy: when writing the `cputune` and it's done with with method now, the current code might not work since the last change.
The second place is writing the policy to the metadata.

If we leave it with `NONE`, we need only two places we need to handle it in a different way - allocating the CPUs and when writing it to the metadata.
Then, the current implementation of this function will work as expected (we are having `NONE` so we will return the dynamic CPU pinning).

Reply:
> If we leave it with `NONE`, we need only two places we need to handle it in a different way - allocating the CPUs and when writing it to the metadata.

`when writing to the metadata` - that's fine. VDSM doesn't really care if the pinning was specified by a user or not - all it cares about is whether this means the pinning is exclusive or not (leaving aside the policies we didn't get to). So I think it's ok to check there and say that if there is pinning defined and the policy is not dedicated, we add MANUAL to the metadata

For the engine it does make a difference because we determine whether to take the pinning from vm_static or vm_dynamic based on that but if we manage to ensure that we always have either manual pinning or dynamic pinning defined and then we check one and fallback to the other if the first doesn't exist, then that's also ok

The only exception is what we discussed offline - returning back to clients policy=NONE and some pinning defined might be surprising but this can be covered with documentation that NONE doesn't mean that there is no pinning, there might be pinning generated by NUMA pinning

Reply:
There is one more option - when running a VM with manual pinning, copy the manual pinning from vm_static into the `currentCpuPinning` in vm_dynamic. That way we could work with just vm_dynamic value throughout the whole code.

Reply:
> There is one more option - when running a VM with manual pinning, copy the manual pinning from vm_static into the `currentCpuPinning` in vm_dynamic. That way we could work with just vm_dynamic value throughout the whole code.

yeah, this approach also crossed my mind - to have cpu pinning taken from the dynamic part and custom cpu pinning in the static part and just copy the one from the static part to the dynamic part rather than checking if the pinning in the dynamic part exists and falling back to the static part if it doesn't

Reply:
... but let's keep that for another day